File: validparentheses.go
// Time Complexity : O(s)
// Space Complexity : O(s)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
have a stack
loop over characters of string
if they are opening then add to stack
else if stack is empty return false else if its not the complementary value of popped element return false

after the loop if stack is not empty return false
return true
*/
package main

import "fmt"

func isValid(s string) bool {
	st := []byte{}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '(' || c == '{' || c == '[' {
			st = append(st, c)
		} else {
			if len(st) > 0 {
				co := st[len(st)-1]
				st = st[:len(st)-1]
				if c == ')' && co != '(' {
					return false
				}
				if c == ']' && co != '[' {
					return false
				}
				if c == '}' && co != '{' {
					return false
				}
			} else {
				return false
			}
		}
	}
	if len(st) > 0 {
		return false
	}
	return true
}

func MainValidParantheses() {
	fmt.Println(isValid("(]")) //expected false
}

File: ValidParentheses/ValidParentheses.java
/* Time Complexity : O(n)
 *   n - length of String - s  */
/* Space Complexity : O(n) 
 * stack size */
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

/*While iterating over the string, if we encounter an open bracket, push it's sorrespondinf closing bracket to stack.
if we encounter a closing bracket, pop from stack and  check if it is same else return false*/

class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        Stack<Character> st = new Stack<>();
        for(int i = 0; i < n; i++){
            char ch = s.charAt(i);
            if(ch == '('){
                st.push(')');
            } else if(ch == '{'){
                st.push('}');
            } else if(ch == '['){
                st.push(']');
            } else if(st.isEmpty() || st.pop() != ch){
                return false;
            }
        }
        if(!st.isEmpty()) return false;
        return true;
    }
}